File: API/PageAndRDL/ReportWizard/C#/ReportWizard/UI/WizardSteps/SelectOutputFields.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Forms;
using GrapeCity.ActiveReports.Samples.ReportWizard.MetaData;

namespace GrapeCity.ActiveReports.Samples.ReportWizard.UI.WizardSteps
{
	public partial class SelectOutputFields : BaseStep
	{
		private class SourcedFieldMetaData
		{
			public static readonly DataFormats.Format Format;

			static SourcedFieldMetaData()
			{
				Format = DataFormats.GetFormat(typeof( SourcedFieldMetaData ).FullName);
			}

			public FieldMetaData Field;

			public object Source;
		}

		private readonly BindingList<FieldMetaData> _availableFields;
		private readonly BindingList<FieldMetaData> _selectedFields;
		private ReportMetaData _report;

		public SelectOutputFields()
		{
			_report = null;
			_availableFields = new BindingList<FieldMetaData>();
			_selectedFields = new BindingList<FieldMetaData>();
			InitializeComponent();
		}

		protected override void OnDisplay(bool firstDisplay)
		{
			if (_report != ReportWizardState.SelectedMasterReport)
			{
				_report = ReportWizardState.SelectedMasterReport;
				LoadFields();
			}
		}

		public override void UpdateState()
		{
			ReportWizardState.DisplayFields.Clear();
			ReportWizardState.DisplayFields.AddRange(_selectedFields);
		}

		private void LoadFields()
		{
			_availableFields.Clear();
			_selectedFields.Clear();
			foreach (KeyValuePair<string, FieldMetaData> keyValuePair in _report.Fields)
			{
				_availableFields.Add(keyValuePair.Value);
			}
			if (availableOutputFields.DataSource == null)
			{
				availableOutputFields.DisplayMember = "Title";
				availableOutputFields.DataSource = _availableFields;
			}
			if (selectedOutputFields.DataSource == null)
			{
				selectedOutputFields.DisplayMember = "Title";
				selectedOutputFields.DataSource = _selectedFields;
			}
		}

		private void availableOutputFields_GetDataObject(object sender, DragDropListBox.DataObjectEventArgs e)
		{
			int index = availableOutputFields.IndexFromPoint(e.MouseLocation);
			if (index < 0)
			{
				e.DataObject = null;
				return;
			}
			SourcedFieldMetaData data = new SourcedFieldMetaData();
			data.Field = _availableFields[index];
			data.Source = availableOutputFields;
			e.DataObject = data;
		}

		private void availableOutputFields_GetDragEffects(object sender, DragDropListBox.DragDropEffectsEventArgs e)
		{
			e.DragDropEffects = DragDropEffects.Move;
		}

		private void selectedOutputFields_DragDrop(object sender, DragEventArgs e)
		{
			SourcedFieldMetaData ddoField =
					(SourcedFieldMetaData) e.Data.GetData(SourcedFieldMetaData.Format.Name);
			if (ddoField != null)
			{
				AddFieldToSelected(ddoField.Field);
			}
		}

		private void AddFieldToSelected(FieldMetaData field)
		{
			_selectedFields.Add(field);
			_availableFields.Remove(field);
		}

		private void RemoveFieldFromSelected(FieldMetaData field)
		{
			_availableFields.Add(field);
			_selectedFields.Remove(field);
		}

		private void selectedOutputFields_DragEnter(object sender, DragEventArgs e)
		{
			SourcedFieldMetaData ddoField =
					(SourcedFieldMetaData) e.Data.GetData(SourcedFieldMetaData.Format.Name);
			if (ddoField != null)
			{
				e.Effect = DragDropEffects.Move;
			}
		}

		private void addOutputField_Click(object sender, EventArgs e)
		{
			FieldMetaData field = availableOutputFields.SelectedItem as FieldMetaData;
			if (field != null)
			{
				AddFieldToSelected(field);
			}
		}

		private void removeOutputField_Click(object sender, EventArgs e)
		{
			FieldMetaData field = selectedOutputFields.SelectedItem as FieldMetaData;
			if (field != null)
			{
				RemoveFieldFromSelected(field);
			}
			removeOutputField.Enabled = (selectedOutputFields.SelectedIndices.Count > 0);
		}

		private void OnSelectedIndexChanged(object sender, EventArgs e)
		{
			addOutputField.Enabled = (availableOutputFields.SelectedIndices.Count > 0);
			removeOutputField.Enabled = (selectedOutputFields.SelectedIndices.Count > 0);
		}
	}
}

File: API/PageAndRDL/ReportWizard/VB.NET/ReportWizard/UI/WizardSteps/SelectOutputFields.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Windows.Forms
Imports GrapeCity.ActiveReports.Samples.ReportWizard.MetaData
Namespace UI.WizardSteps
	Partial Public Class SelectOutputFields
		Inherits BaseStep
		Private Class SourcedFieldMetaData
			Public Shared ReadOnly Format As DataFormats.Format
			Shared Sub New()
				Format = DataFormats.GetFormat(GetType(SourcedFieldMetaData).FullName)
			End Sub
			Public Field As FieldMetaData
			Public Source As Object
		End Class
		Private ReadOnly availableFields As BindingList(Of FieldMetaData)
		Private ReadOnly selectedFields As BindingList(Of FieldMetaData)
		Private report As ReportMetaData
		Public Sub New()
			report = Nothing
			availableFields = New BindingList(Of FieldMetaData)()
			selectedFields = New BindingList(Of FieldMetaData)()
			InitializeComponent()
		End Sub
		Protected Overloads Overrides Sub OnDisplay(ByVal firstDisplay As Boolean)
			If Not report Is ReportWizardState.SelectedMasterReport Then
				report = ReportWizardState.SelectedMasterReport
				LoadFields()
			End If
		End Sub
		Public Overloads Overrides Sub UpdateState()
			ReportWizardState.DisplayFields.Clear()
			ReportWizardState.DisplayFields.AddRange(selectedFields)
		End Sub
		Private Sub LoadFields()
			availableFields.Clear()
			selectedFields.Clear()
			For Each keyValuePair As KeyValuePair(Of String, FieldMetaData) In report.Fields
				availableFields.Add(keyValuePair.Value)
			Next
			If (availableOutputFields.DataSource) Is Nothing Then

				availableOutputFields.DisplayMember = "Title"

				'' availableOutputFields 
				availableOutputFields.DataSource = availableFields
			End If
			If (selectedOutputFields.DataSource) Is Nothing Then
				selectedOutputFields.DisplayMember = "Title"

				selectedOutputFields.DataSource = selectedFields
			End If
		End Sub
		Private Sub availableOutputFields_GetDataObject(ByVal sender As Object, ByVal e As DragDropListBox.DataObjectEventArgs) Handles availableOutputFields.GetDataObject
			Dim index As Integer = availableOutputFields.IndexFromPoint(e.MouseLocation)
			If index < 0 Then
				e.DataObject = Nothing
				Return
			End If
			Dim data As New SourcedFieldMetaData()
			data.Field = availableFields(index)
			data.Source = availableOutputFields
			e.DataObject = data
		End Sub
		Private Sub availableOutputFields_GetDragEffects(ByVal sender As Object, ByVal e As DragDropListBox.DragDropEffectsEventArgs) Handles availableOutputFields.GetDragEffects
			e.DragDropEffects = DragDropEffects.Move
		End Sub
		Private Sub selectedOutputFields_DragDrop(ByVal sender As Object, ByVal e As DragEventArgs) Handles selectedOutputFields.DragDrop
			Dim ddoField As SourcedFieldMetaData = CType(e.Data.GetData(SourcedFieldMetaData.Format.Name), SourcedFieldMetaData)
			If Not (ddoField) Is Nothing Then
				AddFieldToSelected(ddoField.Field)
			End If
		End Sub
		Private Sub AddFieldToSelected(ByVal field As FieldMetaData)
			selectedFields.Add(field)
			availableFields.Remove(field)
		End Sub
		Private Sub RemoveFieldFromSelected(ByVal field As FieldMetaData)
			availableFields.Add(field)
			selectedFields.Remove(field)
		End Sub
		Private Sub selectedOutputFields_DragEnter(ByVal sender As Object, ByVal e As DragEventArgs) Handles selectedOutputFields.DragEnter
			Dim ddoField As SourcedFieldMetaData = CType(e.Data.GetData(SourcedFieldMetaData.Format.Name), SourcedFieldMetaData)
			If Not (ddoField) Is Nothing Then
				e.Effect = DragDropEffects.Move
			End If
		End Sub
		Private Sub addOutputField_Click(ByVal sender As Object, ByVal e As EventArgs) Handles addOutputField.Click
			Dim field As FieldMetaData = TryCast(availableOutputFields.SelectedItem, FieldMetaData)
			If Not (field) Is Nothing Then
				AddFieldToSelected(field)
			End If
		End Sub
		Private Sub removeOutputField_Click(ByVal sender As Object, ByVal e As EventArgs) Handles removeOutputField.Click
			Dim field As FieldMetaData = TryCast(selectedOutputFields.SelectedItem, FieldMetaData)
			If Not (field) Is Nothing Then
				RemoveFieldFromSelected(field)
			End If
			removeOutputField.Enabled = (selectedOutputFields.SelectedIndices.Count > 0)
		End Sub
		Private Sub OnSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles selectedOutputFields.SelectedIndexChanged, availableOutputFields.SelectedIndexChanged
			addOutputField.Enabled = (availableOutputFields.SelectedIndices.Count > 0)
			removeOutputField.Enabled = (selectedOutputFields.SelectedIndices.Count > 0)
		End Sub
	End Class
End Namespace
